문나혜: 섹션 6의 미션 제출

moonnahye opened the pull request:
## #️⃣연관된 이슈

#46

## 📝학습 내용

### 🍀문자열과 타입 안정성
어떤 조건에 따라 결과값을 나타내는 메서드를 만들때, 
```java
public int discount(String grade, int price){
    // 코드
}
```
(예 : 등급에 따라 다른 할인율을 계산하는 메서드)
그 조건을 문자열을 사용하면 여러 문제가 있다.
- 문자열은 오타가 발생할수 있고, 유효하지 않은 값이 입력될 수 있다.
- String 타입 사용시 컴파일 단계에서 오류를 감지 할 수 없다.

**해결법 1 : 문자열 상수 사용하기**
상수는 미리 정의한 변수명을 사용하기 때문에 직접입력하는 것 보단 오타를 줄일수있다.
하지만 여전히 메소드의 파타미터 String 타입은 상수가 아니라도 어떤 문자열을 입력할 수있기때문에 상수가 선언되어있는지 모르는 개발자는 아무 문자열을 입력할 수있다.

### 🍀타입 안전 열거형 패턴
그래서 해결하기 위해 고민해서 나온 결과가 타입 안전 열거형 패턴이다.
```java
public class ClassGrade {
    public static final ClassGrade BASIC = new ClassGrade();
    public static final ClassGrade GOLD = new ClassGrade();
    public static final ClassGrade DIAMOND = new ClassGrade();
}
```
등급을 다루는 클래스를 만들고 
각각 상수마다 새로운 인스턴스를 만들어서 대입한다. 

-> 이 방법도 새로운 객체를 생성해서 넣을 가능성이 있으므로 
외부에서 생성할 수 없도록 기본생성자를 private로 변경하자!

```java
public class ClassGrade {
    public static final ClassGrade BASIC = new ClassGrade();
    public static final ClassGrade GOLD = new ClassGrade();
    public static final ClassGrade DIAMOND = new ClassGrade();

// private 생성자 추가
    private ClassGrade() {}
}
```

->  우리가 생성한 BASIC, GOLD, DIAMOND 상수만 사용할수있게 된다. 

- 타입 안전 열거형 패턴의 장점
  - 우리가 생성한 인스턴스만 사용할수있기때문에 미리 정의된 값만 사용하도록 보장된다.
  - 잘못된 값이라면 컴파일 시점에서 확인할수있다.
- 타입 안전 열거형 패턴의 단점
  - 많은 코드를 작성해야한다.

### 🍀열거형 - Enum Type
-> 자바가 제공하는 타입 안전형 열거타입 

```java
public enum Grade {
    BASIC, GOLD, DIAMOND
}
```
이렇게 간단히 표현할수있다.

- Enum의 장점
  - 유효하지 않는 값이 입력될 가능성이 없다.
  - 코드가 간결하고 명확해진다. 일관성이 보장된다.
  - 새로운 등급을 추가하고싶을때 ENUM에 상수를 추가하기만 하면 된다.
- Enum의 주요 메서드
  - values() : 모든 ENUM상수를 포함하는 배열을 반환
  - valueOf(String name) : 주어진 이름과 일치하는 ENUM상수를 반환한다.
- 열거형 정리
  - 열거형은 java.lang.Enum을 자동 상속받으므로 추가로 다른 클래스 상속이 불가하다.
  - 열거형은 인터페이스를 구현할 수 있다. 
  - 열거형에 추상메서드를 선언하고 구현할 수 있다. 

### 스크린샷 (선택)
![bank실행결과](https://github.com/user-attachments/assets/44f9f3c3-6a2f-4114-b44c-fcf2d1965126)
## 💬리뷰 요구사항

ATM에서 사용하는 거래타입들을 ENUM으로 정의했습니다. 
기능을  나누는 것에 고민을 많이 했는데 그 부분을 중점으로 봐주시면감사하겠습니다.
그리고 수정해야할 부분이 있는지 검토부탁드립니다!

moonnahye committed:
feat. 섹션6 미션 제출

SungbinYang commented on nahye/java-mid1/src/section06/ATM.java:
```diff
@@ -0,0 +1,83 @@
+        int amount;
+        switch (type) {
+            case DEPOSIT:
+                System.out.print("입금할 금액을 입력해주세요: ");
+                amount = input.nextInt();
+                myAccount.deposit(amount);
```
입력받는 값에 대한 유효성 검증이 부족합니다!

SungbinYang commented on nahye/java-mid1/src/section06/ATM.java:
```diff
@@ -0,0 +1,83 @@
+                System.out.println(myAccount);
+                break;
+
+            case EXIT:
+                break;
```
EXIT는 run() 메서드에서 이미 처리되었으므로 processTransaction 메서드에 포함될 필요가 없습니다. case EXIT를 제거해 불필요한 로직을 정리할 수 있지 않을까요?

moonnahye replied:
넵 지우겠습니다!

SungbinYang commented on nahye/java-mid1/src/section06/ATM.java:
```diff
@@ -0,0 +1,83 @@
+            case TRANSFER:
+                System.out.print("송금할 통장의 번호를 입력해주세요: ");
+                int accountNumberToTransfer = input.nextInt();
+                if (accountNumberToTransfer == myAccount.getAccountNumber()) {
+                    throw new IllegalArgumentException("송금할 통장은 본인 통장과 달라야합니다.");
```
TRANSFER 로직에서 송금 대상 통장 번호가 동일한 경우 예외를 던지고 있지만, 금액 유효성 검증이 부족합니다.

moonnahye replied:
추가하겠습니다!

SungbinYang commented on nahye/java-mid1/src/section06/ATM.java:
```diff
@@ -0,0 +1,83 @@
+
+    private void processTransaction(Account myAccount, TransactionType type) {
+        int amount;
+        switch (type) {
```
각 타입에 대한 내용들을 메서드로 분리해보심도 좋을 것 같아요!

moonnahye replied:
넵 알겠습니다!

SungbinYang commented on nahye/java-mid1/src/section06/Account.java:
```diff
@@ -0,0 +1,58 @@
+        if (amount <= 0) {
+            throw new IllegalArgumentException("입금액은 0보다 커야합니다.");
+        }
+        if (amount > balance) {
+            throw new IllegalArgumentException("잔액이 모자랍니다.");
+        }
```
입금하는 로직에서 이 부분이 필요할까요?

moonnahye replied:
앗 지우겠습니다!

SungbinYang commented on nahye/java-mid1/src/section06/Account.java:
```diff
@@ -0,0 +1,58 @@
+    }
+
+    public void deposit(int amount) {
+        if (amount <= 0) {
+            throw new IllegalArgumentException("입금액은 0보다 커야합니다.");
+        }
```
일반 ATM기 보면 입금액은 0보다 커야한다는 문구는 나오지 않겠죠? 조금 더 사용자 친화적 문구로 변경해보심도 좋을 것 같아요!

SungbinYang commented on nahye/java-mid1/src/section06/Account.java:
```diff
@@ -0,0 +1,58 @@
+            throw new IllegalArgumentException("잔액이 모자랍니다.");
+        }
+        balance -= amount;
+        targetAccount.balance += balance;
```
transfer 메서드에서 송금 후 대상 계좌의 잔액을 현재 계좌의 잔액으로 잘못 추가하고 있습니다. balance가 아니라 amount를 사용해야 하지 않을까요?

SungbinYang commented on nahye/java-mid1/src/section06/Account.java:
```diff
@@ -0,0 +1,58 @@
+        this.balance = balance;
+    }
+
+    public void deposit(int amount) {
```
송금 출금 입금등 각각 동시성 문제도 고려해보심이 좋을 것 같습니다!

SungbinYang commented on nahye/java-mid1/src/section06/Bank.java:
```diff
@@ -0,0 +1,52 @@
+        createAccount(2222, findCustomerByName("LEE"), 10000);
+    }
+
+    public Customer addNewCustomer(String name, int phoneNumber) {
```
전화번호를 int로 하는게 맞을까요?

moonnahye replied:
수정하겠습니다!

SungbinYang commented on nahye/java-mid1/src/section06/Account.java:
```diff
@@ -0,0 +1,58 @@
+
+public class Account {
+
+    private int accountNumber;
```
계좌번호를 int로 두는게 맞을까요?

SungbinYang commented on nahye/java-mid1/src/section06/Account.java:
```diff
@@ -0,0 +1,58 @@
+    private int accountNumber;
+    private Customer owner;
+
+    private int balance;
```
잔액이 int의 최고범위를 넘어설 수 있지 않을까요?

SungbinYang commented on nahye/java-mid1/src/section06/Bank.java:
```diff
@@ -0,0 +1,52 @@
+    }
+
+    public Customer addNewCustomer(String name, int phoneNumber) {
+        Customer newCustomer = new Customer(name, phoneNumber);
+        customers.add(newCustomer);
```
고객 추가 시 중복 이름과 전화번호에 대한 처리가 없습니다.

SungbinYang commented on nahye/java-mid1/src/section06/Customer.java:
```diff
@@ -0,0 +1,15 @@
+
+public class Customer {
+    private String name;
+    private int phoneNumber;
```
전화번호를 int로 하는게 맞을까요?

SungbinYang reviewed (commented):
코드리뷰 남겨드렸습니다! 확인바랍니다~

moonnahye committed:
fix. 코드리뷰 반영 by 성빈님

- 계좌번호, 전화번호 String으로 타입변경
- 잔액과, 금액관련 변수 Long으로 타입변경
- 거래할 금액에 대해 검증 및 처리하는 processForAmount() 메소드 추가
- Customer에 ID 필드 추가
- addNewCustomer()에 id가 같을경우 예외처리 추가, 전화번호가 같을 경우 예외처리 추가
-

ljh468 reviewed (approved):
위에서 꼼꼼하게 리뷰 해주셨네요. 고생하셨어요 LGTM :)

Darimi2222 commented on nahye/java-mid1/src/section06/Account.java:
```diff
@@ -0,0 +1,55 @@
+    private String accountNumber;
+    private Customer owner;
+
+    private Long balance;
```
금액을 래퍼 클래스로 설정하신 이유가 있을까요?

moonnahye replied:
오잉 long으로 수정하는게 나을것같아요! 수정하겠습니다

Darimi2222 commented on nahye/java-mid1/src/section06/Account.java:
```diff
@@ -0,0 +1,55 @@
+        this.balance = balance;
+    }
+
+    public void deposit(Long amount) {
```
이런 매개변수도 래퍼로 하신 이유가 있을까요? 래퍼로 하신 명확한 사유가 존재한다면 매개변수에 null이 들어올 때의 처리도 필요해보여요!

Darimi2222 commented on nahye/java-mid1/src/section06/ATM.java:
```diff
@@ -0,0 +1,105 @@
+
+                TransactionType type = TransactionType.findByDescription(selectType);
+
+                if (type == TransactionType.EXIT) {
```
종료 부분만 if문 밖에서 처리하는 게 조금 어색한 것 같기도 해요!

moonnahye replied:
수정해보도록 하겠습니다

Darimi2222 reviewed (commented):
코멘트 몇 개 남기고 갑니다!
오늘도 수고 많으셨습니다. :)

hmkim199 reviewed (approved):
사용자 입력을 받도록 상세히 구현하신 부분 잘 봤습니다!! 고생 많으셨습니다.

sm9171 commented on nahye/java-mid1/src/section06/ATM.java:
```diff
@@ -0,0 +1,105 @@
+        this.bank = bank;
+    }
+
+    public void run() {
```
오 사용자 인터페이스도 구현해주셨군요..!

sm9171 commented on nahye/java-mid1/src/section06/TransactionType.java:
```diff
@@ -0,0 +1,33 @@
+        return description;
+    }
+
+    public static TransactionType findByDescription(String description) {
```
enum객체에 find메소드 구현해주셨군요!

sm9171 reviewed (approved):
전체적으로 완성도 있게 구현해주신것 같아요.
수고하셨습니다.

moonnahye committed:
fix. 코드리뷰 반영 by 채원님

- 잔액과, 금액관련 변수 long 타입으로 변경
- 종료하는 기능을 ATM run()메서드에서 제거 후 processTransaction()에 적용

SungbinYang reviewed (approved):
수정하신 부분 확인했습니다!

tdronnie commented on nahye/java-mid1/src/section06/ATM.java:
```diff
@@ -0,0 +1,99 @@
+                Account myAccount = bank.findAccountByNumber(accountNumber);
+
+                System.out.print("원하시는 거래를 입력해주세요 >> ");
+                for (TransactionType type : TransactionType.values()) {
+                    System.out.print(type.getDescription() + ", ");
+                }
```
Enum 관련 메서드를 활용하려하신게 보이네요!

tdronnie commented on nahye/java-mid1/src/section06/ATM.java:
```diff
@@ -0,0 +1,99 @@
+    private void processTransaction(Account myAccount, TransactionType type) {
+        long amount;
+        try {
+            switch (type) {
```
switch문에 Enum을 활용하셨군요!
case 문 안에 많은 로직이 들어가 있어서 메서드추출을 고려해봐도 좋다고 생각해요

moonnahye replied:
넵 추출해보겠습니다!

tdronnie commented on nahye/java-mid1/src/section06/Bank.java:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    public Customer addNewCustomer(String id, String name, String phoneNumber) {
+        for (Customer customer : customers) {
```
새로운 고객정보를 저장할 때 항상 모든 고객정보를 조회하는 것보다 Map과 같은 자료구조로 조회를 O(1)로   하는 것도 방법일 것 같습니다!

moonnahye replied:
이것도 한번 수정해보겠습니다!

tdronnie commented on nahye/java-mid1/src/section06/TransactionType.java:
```diff
@@ -0,0 +1,33 @@
+package section06;
+
+public enum TransactionType {
```
Enum클래스에 ATM의 각 거래 로직을 어느정도 위임해줄 수도 있을 것 같아요!

tdronnie reviewed (approved):
몇가지 코멘트 남겨드렸습니다! 사소한 리팩토링적인 사항이기 때문에 approve하겠습니다!

moonnahye committed:
fix. 코드리뷰 반영 by 세은님

- ATM의 processTransaction() switch문의 로직 메서드로 추출

moonnahye committed:
fix. 코드리뷰 반영 by 세은님

- Bank클래스의 accouts와 customers 필드를 List 자료형에서 Map 자료형으로 변경

Darimi2222 reviewed (approved):
수정하신 부분 확인했습니다! 수고 많으셨습니다!
